File: lang/python/test/muse/data_structures/CircularlyLinkedListTest.py
from muse.data_structures.CircularlyLinkedList import CircularlyLinkedList
from muse.util import TestRunner

def testCircularlyLinkedList() -> bool:
    size = 8192

    list = CircularlyLinkedList()

    for i in range(1, size + 1):
        list.append(i)

    if list.size() != size:
        return False

    for i in range(size):
        if list.get(i) != i + 1:
            return False

    for i in range(size):
        list.set(i, size - i)

    for i in range(size):
        if list.get(i) != size - i:
            return False

    x = 0
    y = 0

    i, j = 0, size - 1
    while i < j:
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
        i, j = i + 1, j - 1

    for i in range(size):
        if list.get(i) != i + 1:
            return False

    for i in range(size, 0, -1):
        if list.back() != i:
            return False
        list.eject()

    return list.isEmpty()

def main() -> None:
    TestRunner.parseTest(testCircularlyLinkedList())

if __name__ == "__main__":
    main()

File: nodejs/muse/data_structures/CircularlyLinkedListTest.js
'use strict'

const CircularlyLinkedList = require('muse/data_structures/CircularlyLinkedList')
const TestRunner = require('muse/util/TestRunner')

const testCircularlyLinkedList = () => {
  const size = 8192

  const list = new CircularlyLinkedList.CircularlyLinkedList()

  for (let i = 1; i <= size; i++) {
    list.append(i)
  }

  if (list.size() !== size) {
    return false
  }

  for (let i = 0; i < size; i++) {
    if (list.get(i) !== i + 1) {
      return false
    }
  }

  for (let i = 0; i < size; i++) {
    list.set(i, size - i)
  }

  for (let i = 0; i < size; i++) {
    if (list.get(i) !== size - i) {
      return false
    }
  }

  for (let i = 0, j = size - 1; i < j; i++, j--) {
    const x = list.get(i)
    const y = list.get(j)

    list.remove(i)
    list.insert(i, y)
    list.remove(j)
    list.insert(j, x)
  }

  for (let i = 0; i < size; i++) {
    if (list.get(i) !== i + 1) {
      return false
    }
  }

  for (let i = size; i >= 1; i--) {
    if (list.back() !== i) {
      return false
    }
    list.eject()
  }

  return list.isEmpty()
}

if (module === require.main) {
  TestRunner.pick(testCircularlyLinkedList)
}
